### PR TITLE
Fix all amp-ad related type errors and turn on type checking.

### PR DESCRIPTION
**Pull request by cramforce**
CC @ampproject/a4a

### CODE REVIEW

File: ads/google/a4a/url-builder.js [hunk 1]
```diff
@@ -14,8 +14,8 @@
  * limitations under the License.
  */
 
-/** typedef {{name: string, value: ?(string|number)}} */
-let QueryParameterDef;
+/** @typedef {{name: string, value: (string|number|null)}} */
```
Comment (lannka) [inline ads/google/a4a/url-builder.js]: oh, `?(string|number)` doesn't work? how about `(?string|?number)`?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 2]
```diff
@@ -305,7 +306,7 @@ export class AmpA4A extends AMP.BaseElement {
       /** @return {!Promise<?string>} */
       .then(() => {
         checkStillCurrent(promiseId);
-        return this.getAdUrl();
+        return /** @type {!Promise<?string>} */ (this.getAdUrl());
```
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: is this needed? isn't it a duplication of the annotation above `then()`?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -538,17 +539,17 @@ export class AmpA4A extends AMP.BaseElement {
...
+   * @return {!Promise<?ArrayBuffer>}  Promise to a guaranteed-valid AMP creative
    *   or null if the creative is unsigned or invalid.
    * @private
    */
   validateAdResponse_(creative, signature) {
     // Validate when we have a signature and we have native crypto.
     if (!signature) {
       // Guaranteed not a AMP creative.
-      return Promise.resolve(null);
+      return /** @type {!Promise<?ArrayBuffer>} */ (Promise.resolve(null));
```
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: same here, why is this necessary?

File: ads/google/a4a/url-builder.js (hunk 1 above)
Comment (cramforce) [inline ads/google/a4a/url-builder.js]: I find this clearer when making union with non-nullable-by-default-types.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (cramforce) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Same reason. Generally the compiler cannot do type math with generic and hence needs a lot of casts for API that aren't designed to avoid that.

Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: I can understand the previous case, but here should be straight forward to the compiler right? It's a Promise of null, why compiler think it doesn't match the return type?

Comment (cramforce) [inline extensions/amp-a4a/0.1/amp-a4a.js]: As I said above: The compiler cannot do type math with generics. For this reason while

'null' is an acceptable value where 'string|null' is expected, the compiler will not accept `Container<null>` where `Container<string|null>` is expected. I'd ask the compiler team if they intend to fix this.

Comment (erwinmombay) [inline extensions/amp-a4a/0.1/amp-a4a.js]: you could also do `{!Promise<!ArrayBuffer>|!Promise<null>}` right?

Comment (cramforce) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Yeah, but we don't do that.

Its really sad. Basically one shouldn't use type unions on generics. I want to keep these type fix CLs small. Owners of the code can change it to whatever they want.
